nni/nas/execution/engine.py

```diff
@@ -0,0 +1,313 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+import logging
+import time
+from collections import defaultdict
+from typing import Any, Iterable, NewType, Callable, Type, overload
+
+from nni.nas.space import ExecutableModelSpace, ModelStatus
+
+from .event import ModelEventCallbacks, ModelEvent, ModelEventType, FinalMetricEvent, IntermediateMetricEvent, TrainingEndEvent
+
+__all__ = [
+    'WorkerInfo', 'ExecutionEngine', 'Middleware',
+]
+
+
+WorkerInfo: Type[Any] = NewType('WorkerInfo', Any)
+"""
+Won't be needed in short-term. Left from v2.x era.
+
+This describes the properties of a worker machine. (e.g. memory size)
+"""
+
+_logger = logging.getLogger(__name__)
+
+
+class ExecutionEngine:
+    """
+    The abstract interface of execution engine.
+
+    Execution engine is responsible for *executing* the submitted models.
+    The engine has the freedom to choose the execution environment. For example,
+    whether to execute it instantly in the current process, or send it to NNI training service (e.g., local / remote).
+    It may also optimize the workloads with techniques like CSE, or even doing benchmark queries.
+
+    Note that some engines might reply on certain model space formats.
+    For example, some engines might require the model space to be a graph, to do certain optimizations.
+
+    Every subclass of class:`ExecutableModelSpace` has its general logic (i.e., code) of execution
+    defined in its class. But the interpretation of the logic depends on the engine itself.
+
+    In synchronized use case, the strategy will have a loop to call `submit_models` and `wait_models` repeatedly,
+    and will receive metrics from `ExecutableModelSpace` attributes.
+    Execution engine could assume that strategy will only submit graph when there are available resources (for now).
+
+    In asynchronized use case, the strategy will register a listener to receive events,
+    while still using `submit_models` to train.
+
+    There might be some util functions benefit all optimizing methods,
+    but non-mandatory utils should not be covered in abstract interface.
+    """
+
+    def __init__(self) -> None:
+        self._callbacks: ModelEventCallbacks = defaultdict(list)
+
+    def __repr__(self) -> str:
+        return f'{self.__class__.__name__}({self.extra_repr()})'
+
+    def extra_repr(self) -> str:
```

Comment:
Is this used anywhere other than `__repr__()`?
If no, it looks unnecessary and harms readability.

Reply:
Subclass might override `extra_repr()`, saving the trouble of rewriting whole `__repr__`.